API-strategie-extensies/ext-security.md

```diff
@@ -15,16 +15,21 @@ One should secure all APIs assuming they can be accessed from any location on th
 
 <div class="rule" id="api-11">
   <p class="rulelab"><strong>API-11</strong>: Secure connections using TLS</p>
+  <p>Statement:</p>
   <p>Secure connections using TLS following the latest NCSC guidelines [[NCSC.TLS]].</p>
+  <p>Implications:</p>
   <p>Since the connection is always secured, the access method can be straightforward. This allows the application of basic access tokens instead of encrypted access tokens.</p>
 </div>
 
-Even when using TLS-based secured connections information in URIs is not secured. URIs can be cached and logged outside of the servers controlled by clients and servers. Any information contained in them should therfor be considered readable by anyone with access to the netwerk being used (in case of the internet the whole world) and MUST NOT contain any sensitive information. Neither client secrets used for authentication, privacy sensitive informations suchs as BSNs nor any other information which should not be shared. Be aware that queries (anything after the '?' in a URI) are also part of an URI.
-
 <div class="rule" id="api-58">
   <p class="rulelab"><strong>API-58</strong>: No sensitive information in URIs</p>
-  <p>Do not put any sensitive information in URIs</p>
-  <p>Even when the connection is secure URIs can be cached and logged, in systems outside the control of client and/or server.</p>
+  <p>Statement:</p>
+  <p>Do not put any sensitive information in URIs - Use the POST or QUERY HTTP operation instead.</p>
+  <p>Rationale:</p>
+  <p>Even when using TLS-based secured connections information in URIs is not secured. URIs can be cached and logged outside of the servers controlled by clients and servers. Any information contained in them should therefore be considered readable by anyone with access to the network being used (in case of the internet the whole world) and MUST NOT contain any sensitive information. Neither client secrets used for authentication, privacy sensitive information such as BSNs nor any other information which should not be shared. Be aware that queries (anything after the '?' in a URI) are also part of an URI.</p>
```

Comment:
This is not correct, URI's are encrypted the same way the request and body are encrypted. Only the domain name (part of the URI) will be used to set up the TLS connection, thus will be available in plain text.  Everything after (and before) the domain name will be encrypted, thus only available for organisations / clients that van decrypt the TLS connection.

More in general, this advice would have a large impact when followed, so I'd like to discuss that in more detail. Or did I miss the discussion?

Reply:
in addition https://datatracker.ietf.org/doc/html/draft-ietf-httpbis-safe-method-w-body-03 offers another alternative to explore that has not been explored in previous discussions

Reply:
#219

Reply:
Hmmm, design rule 58 says: 

```
Do not put any sensitive information in URIs

Even when the connection is secure URIs can be cached and logged, in systems outside the control of client and/or server.
```

Following the same logic you can say:

```
Do not put any sensitive information in _http message bodies_

Even when the connection is secure, _http message bodies_ can be cached and logged, in systems outside the control of client and/or server.
```

They are encrypted with the same certificate when using TLS. Also, logging of body is not obscure, it is more and more common thanks to extra  security features (although mostly it's truncated to the first 1024 characters of that body).

In my opinion this Design Rule 58 needs to be revisited as there is more to it than this rule suggests

Reply:
The intent of this change was to suggest a solution for the rule that now only states to 'not do something'. In my opinion the rule should at least suggest what one should or must do. Hence this suggested change.

Reply:
I think it's too early to mention the `QUERY` operation, since this draft is still in a very early stage. There is hardly any support in the current software landscape.

Reply:
> The intent of this change was to suggest a solution for the rule that now only states to 'not do something'. In my opinion the rule should at least suggest what one should or must do. Hence this suggested change.

I'd agree but by reading the proposed solution I conclude that the rationale of the rule itself is not valid. Everywhere where the query parameters are available in plaintext, the body will be available as well. And the response, most likely containing sensitive information as well, will be available at those points as well. The discussion should be about if there is actually a problem to be solved here, and how big that problem is compared to deviation from the technical standards.

Reply:
I agree @ehotting.
As discussed today in the working group we will rewrite the rule to ben more informative about the risks involved when sharing sensitive information with various client types (as described in the [oauth nl profile](https://publicatie.centrumvoorstandaarden.nl/api/oauth/#client-types))